### PR TITLE
Refactor DisplayPackageInfo: `temporaryPackageInfo`

### PR DESCRIPTION
**Pull request by lennartclaas**
### Summary of changes

Apply the refactoring regarding `DisplayPackageInfo` to `temporaryPackageInfo`. This change entails changes to many components that get or set `temporaryPackageInfo`.

### Context and reason for change

Improve separation of display and business logic by introducing `DisplayPackageInfo` in all components that display attributions.

### CODE REVIEW

File: src/Frontend/state/selectors/all-views-resource-selectors.ts [hunk 1]
```diff
@@ -130,18 +129,24 @@ export function getCurrentAttributionId(state: State): string | null {
...
+): DisplayPackageInfo | null {
   const selectedAttributionId = getSelectedAttributionId(state);
 
   if (!selectedAttributionId) {
     return null;
   }
   const attributions = getManualAttributions(state);
 
-  return attributions[selectedAttributionId];
+  // TODO: Will this retrigger renderings if the state has not changed? (New object is created with convert function)
```
Comment (benedikt-richter) [inline src/Frontend/state/selectors/all-views-resource-selectors.ts:140]: When will this comment be resolved?

File: src/Frontend/util/convert-package-info.ts [hunk 2]
```diff
@@ -5,13 +5,17 @@
...
 import { getPackageInfoKeys } from '../../shared/shared-util';
+import { EMPTY_DISPLAY_PACKAGE_INFO } from '../shared-constants';
 import { getDisplayPackageInfoKeys } from './get-display-package-info-keys';
 import { shouldNotBeCalled } from './should-not-be-called';
 
 export function convertPackageInfoToDisplayPackageInfo(
-  packageInfo: PackageInfo,
-  attributionIds: Array<string>
+  packageInfo?: PackageInfo | null,
+  attributionIds?: Array<string>
```
Comment (benedikt-richter) [inline src/Frontend/util/convert-package-info.ts:14]: Why are these arguments optional?
ALso, what is the use case of converting `null`?

File: src/Frontend/util/get-stripped-package-info.ts [hunk 3]
```diff
@@ -4,23 +4,44 @@
...
-  return removeExcessProperties(strippedTemporaryPackageInfo);
+  return removeExcessProperties(strippedPackageInfo);
 }
 
 function removeExcessProperties(rawPackageInfo: PackageInfo): PackageInfo {
   const packageInfoKeys = getPackageInfoKeys();
   return pick(rawPackageInfo, packageInfoKeys);
 }
+
+export function getStrippedDisplayPackageInfo(
```
Comment (benedikt-richter) [inline src/Frontend/util/get-stripped-package-info.ts:27]: This function is used to calculate a variable `temporaryPackageInfoWithoutConfidence`. This seems not to fit the actual implementation

Comment (benedikt-richter) [reply src/Frontend/util/get-stripped-package-info.ts:27]: This function is untested

**Review by benedikt-richter — CHANGES_REQUESTED** [review]

File: src/Frontend/Components/AttributionColumn/AttributionColumn.tsx [hunk 4]
```diff
@@ -174,8 +175,8 @@ export function AttributionColumn(props: AttributionColumnProps): ReactElement {
       buttonText: ButtonText.Undo,
       disabled: !packageInfoWereModified,
       onClick: (): void => {
-        updatePurl(initialPackageInfo);
-        dispatch(setTemporaryPackageInfo(initialPackageInfo));
+        updatePurl(initialDisplayPackageInfo);
```
[the remark this answers is not in the document]
Comment (lennartclaas) [reply src/Frontend/Components/AttributionColumn/AttributionColumn.tsx:178]: Will be done in next PR.

File: src/Frontend/Components/ReplaceAttributionPopup/ReplaceAttributionPopup.tsx [hunk 5]
```diff
@@ -59,11 +62,22 @@ export function ReplaceAttributionPopup(): ReactElement {
...
+
   function getPackageCard(attributionId: string): ReactElement {
-    const displayPackageInfo = convertPackageInfoToDisplayPackageInfo(
-      attributions[attributionId],
-      [attributionId]
-    );
+    const displayPackageInfo: DisplayPackageInfo =
+      getAttributionFromDisplayAttributionsWithCount(
+        attributionId,
+        displayAttributions
```
[the remark this answers is not in the document]
Comment (lennartclaas) [reply src/Frontend/Components/ReplaceAttributionPopup/ReplaceAttributionPopup.tsx:79]: I thought it was cleaner (according to our principal 'DisplayPackageInfo as early as possible'. But in terms of performance, it does not make sense, I guess. I reverted the changes.

File: src/Frontend/state/actions/resource-actions/__tests__/save-actions.test.ts [hunk 6]
```diff
@@ -209,8 +216,12 @@ describe('The savePackageInfo action', () => {
...
       savePackageInfo('/root/src/', null, testTemporaryPackageInfo)
     );
-    expect(getPackageInfoOfSelected(testStore.getState())).toEqual(
-      expectedTemporaryPackageInfo
+    expect(
+      convertDisplayPackageInfoToPackageInfo(
+        getDisplayPackageInfoOfSelected(testStore.getState())
+      )
+    ).toEqual(
+      convertDisplayPackageInfoToPackageInfo(expectedTemporaryPackageInfo)
```
[the remark this answers is not in the document]
Comment (lennartclaas) [reply src/Frontend/state/actions/resource-actions/__tests__/save-actions.test.ts:224]: It's to ignore random uuids as attributioIds, but I changed it to a clearer version.

File: src/Frontend/state/selectors/all-views-resource-selectors.ts (hunk 1 above)
Comment (lennartclaas) [reply src/Frontend/state/selectors/all-views-resource-selectors.ts:140]: Can you test with real data whether this leads to performance drops?

File: src/Frontend/state/selectors/audit-view-resource-selectors.ts [hunk 7]
```diff
@@ -108,12 +109,16 @@ export function getAttributionIdOfDisplayedPackageInManualPanel(
...
 export function getAttributionOfDisplayedPackageInManualPanel(
   state: State
-): PackageInfo | null {
+): DisplayPackageInfo | null {
   const attributionId: string | null =
     getAttributionIdOfDisplayedPackageInManualPanel(state);
   if (attributionId) {
     const manualAttributions: Attributions = getManualAttributions(state);
-    return manualAttributions[attributionId];
+    // TODO: Will this retrigger renderings if the state has not changed? (New object is created with convert function)
```
[the remark this answers is not in the document]
Comment (lennartclaas) [reply src/Frontend/state/selectors/audit-view-resource-selectors.ts:117]: See above.

File: src/Frontend/util/convert-package-info.ts (hunk 2 above)
Comment (lennartclaas) [reply src/Frontend/util/convert-package-info.ts:14]: For making things easier in other parts of the code. However, I reverted the changes to make things clearer.

File: src/Frontend/util/convert-package-info.ts [hunk 8]
```diff
@@ -70,8 +74,11 @@ export function convertPackageInfoToDisplayPackageInfo(
 }
 
 export function convertDisplayPackageInfoToPackageInfo(
-  displayPackageInfo: DisplayPackageInfo
+  displayPackageInfo?: DisplayPackageInfo | null
```
[the remark this answers is not in the document]
Comment (lennartclaas) [reply src/Frontend/util/convert-package-info.ts:77]: See above.

**Comment by benedikt-richter** [comment]
In attributionColumn the following code still has to be adapted:
```
  const attributionIdsToResolveOrUnresolve = isDisplayPackageInfo(
    temporaryPackageInfo
  )
    ? temporaryPackageInfo.attributionIds
    : []; // TODO: remove when refactoring

```

**Comment by benedikt-richter** [comment]
The subpanels of the AttributionColumn still expect `PackageInfo`. This should be ``DisplayPackageInfo`.

**Review by benedikt-richter — APPROVED** [review]

**Pull request merged**